chain/index/events.go

```diff
@@ -0,0 +1,583 @@
+			// of which event we are dealing with and create a new one each time we see a new id
+			if row.id != currentID {
+				if ce != nil {
+					ces = append(ces, ce)
+					ce = nil
+					// Unfortunately we can't easily incorporate the max results limit into the query due to the
+					// unpredictable number of rows caused by joins
+					// Break here to stop collecting rows
+					if f.MaxResults > 0 && len(ces) >= f.MaxResults {
+						break
+					}
+				}
+
+				currentID = row.id
+				ce = &CollectedEvent{
+					EventIdx: row.eventIndex,
+					Reverted: row.reverted,
+					Height:   abi.ChainEpoch(row.height),
+					MsgIdx:   row.messageIndex,
+				}
+
+				if row.emitterAddr == nil {
+					ce.EmitterAddr, err = address.NewIDAddress(row.emitterID)
+					if err != nil {
+						return nil, xerrors.Errorf("failed to parse emitter id: %w", err)
+					}
+				} else {
+					ce.EmitterAddr, err = address.NewFromBytes(row.emitterAddr)
+					if err != nil {
+						return nil, xerrors.Errorf("parse emitter addr: %w", err)
+					}
+				}
+
+				tsKeyCid, err := cid.Cast(row.tipsetKeyCid)
+				if err != nil {
+					return nil, xerrors.Errorf("parse tipsetkey cid: %w", err)
+				}
+
+				ts, err := si.cs.GetTipSetByCid(ctx, tsKeyCid)
+				if err != nil {
+					return nil, xerrors.Errorf("get tipset by cid: %w", err)
+				}
+				if ts == nil {
+					return nil, xerrors.Errorf("failed to get tipset from cid: tipset is nil for cid: %s", tsKeyCid)
+				}
+
+				ce.TipSetKey = ts.Key()
+
+				ce.MsgCid, err = cid.Cast(row.messageCid)
+				if err != nil {
+					return nil, xerrors.Errorf("parse message cid: %w", err)
+				}
+			}
+
+			ce.Entries = append(ce.Entries, types.EventEntry{
+				Flags: row.flags[0],
+				Key:   row.key,
+				Codec: row.codec,
+				Value: row.value,
+			})
+		}
+
+		if ce != nil {
+			ces = append(ces, ce)
+		}
+
+		if len(ces) == 0 {
+			return nil, nil
+		}
+
+		// collected event list is in inverted order since we selected only the most recent events
+		// sort it into height order
+		sort.Slice(ces, func(i, j int) bool { return ces[i].Height < ces[j].Height })
+
+		return ces, nil
+	}
+
+	values, query, err := makePrefillFilterQuery(f)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to make prefill filter query: %w", err)
+	}
+
+	stmt, err := si.db.Prepare(query)
+	if err != nil {
+		return nil, xerrors.Errorf("prepare prefill query: %w", err)
+	}
+	defer func() { _ = stmt.Close() }()
+
+	ces, err := getEventsFnc(stmt, values)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get events: %w", err)
+	}
+	if len(ces) == 0 {
+		if f.MaxHeight > 0 {
+			head := si.cs.GetHeaviestTipSet()
+			if head == nil {
+				return nil, errors.New("failed to get head: head is nil")
+			}
+			headHeight := head.Height()
+			maxLookBackHeight := headHeight - maxLookBackForWait
+
+			if f.MaxHeight <= maxLookBackHeight {
+				return nil, si.checkTipsetIndexedStatus(ctx, f)
+			}
+		}
```

Comment:
There's another case that matters here: `f.TipsetCid != cid.Undef`. If they supply an ancient tipset and it has no events then we shouldn't bother waiting. In that case we could probably afford to do a tipset lookup in the chainstore, get t he height and use that in the same way that we're using `MaxHeight` here.

```suggestion
		height := f.MaxHeight
		if f.TipsetCid != cid.Undef {
			ts, err := si.cs.GetTipSetByCid(ctx, f.TipsetCid)
			if err != nil {
				return nil, xerrors.Errorf("failed to get tipset by cid: %w", err)
			}
			if ts == nil {
				return nil, xerrors.Errorf("failed to get tipset from cid: tipset is nil for cid: %s", f.TipsetCid)
			}
			height = ts.Height()
		}
		if height > 0 {
			head := si.cs.GetHeaviestTipSet()
			if head == nil {
				return nil, errors.New("failed to get head: head is nil")
			}
			headHeight := head.Height()
			maxLookBackHeight := headHeight - maxLookBackForWait

			// if the height is old enough, we'll assume the index is caught up to it and not bother
			// waiting for it to be indexed
			if height <= maxLookBackHeight {
				return nil, si.checkTipsetIndexedStatus(ctx, f)
			}
		}
```

Reply:
I think the `ts == nil` check is unnecessary in here but you have it elsewhere in this file so maybe you're relying on that for your tests where you override the chainstore?